library/psa_crypto.c

```diff
@@ -710,8 +773,26 @@ static psa_algorithm_t psa_key_policy_algorithm_intersection(
         ( PSA_ALG_FULL_LENGTH_MAC( alg1 ) ==
           PSA_ALG_FULL_LENGTH_MAC( alg2 ) ) )
     {
-        size_t alg1_len = PSA_MAC_TRUNCATED_LENGTH( alg1 );
-        size_t alg2_len = PSA_MAC_TRUNCATED_LENGTH( alg2 );
+        /* Calculate the actual minimal output length for both sides */
+        size_t alg1_len = 0;
+        size_t alg2_len = 0;
+        if( PSA_SUCCESS != psa_get_mac_output_length(
+                                PSA_ALG_TRUNCATED_MAC( alg1,
+                                    PSA_MAC_TRUNCATED_LENGTH( alg1 ) ),
```

Comment:
`alg1` could be a wildcard, in which case the `psa_get_mac_output_length` function will return an error.

Reply:
I extended the comment above to make that clearer.

Reply:
I see. I don't get that from the extended comment at all. `alg1_len` and `alg2_len` aren't actaully “requested” output lengths: they're policy elements. It happens that the structure of MAC algorithm policies can be decomposed into three parts: the base algorithm, a length, and a two-valued modifier “exact length” or “minimum length”. `PSA_ALG_TRUNCATED_MAC( alg, PSA_MAC_TRUNCATED_LENGTH( alg ) )` takes such a policy and forces the length modifier to “exact length”. `alg & ~PSA_ALG_MAC_AT_LEAST_THIS_LENGTH_FLAG` would be clearer.

Since the code depends on what policies are possible anyway, using the direct implementation (`& ~FLAG`) rather than relying on the abstractions wouldn't make the code more fragile. So I'd prefer to do that. I'm ok with going through the abstraction, but only if it's clear what's going on, preferably on an auxiliary function or macro with a comprehensible name.

Reply:
Clarified (and reworked)